server/market/market.go

```diff
@@ -3,30 +3,251 @@
+	return bookUpdates
+}
+
+func (m *Market) SubmitOrder(ord order.Order) error {
+	return <-m.SubmitOrderAsync(ord)
+}
+
+func (m *Market) SubmitOrderAsync(ord order.Order) <-chan error {
+	sig := newOrderUpdateSignal(ord)
+	m.orderRouter <- sig
+	return sig.errChan
+}
+
+func NewMarket(ctx context.Context, lotSize, epochDuration uint64, baseAsset, quoteAsset *asset.Asset,
+	storage db.DEXArchivist, authMgr swap.AuthManager) (*Market, error) {
+
+	mktInfo, err := types.NewMarketInfoFromSymbols(baseAsset.Symbol, quoteAsset.Symbol, lotSize, epochDuration)
+	if err != nil {
+		return nil, err
+	}
+
+	swapperCfg := &swap.Config{
+		Ctx: ctx,
+		Assets: map[uint32]*asset.Asset{
+			mktInfo.Base:  baseAsset,
+			mktInfo.Quote: quoteAsset,
+		},
+		Storage:          storage,
+		AuthManager:      authMgr,
+		BroadcastTimeout: 10 * time.Second, // TODO: who sets this?
+	}
+
+	return &Market{
+		ctx:           ctx,
+		marketInfo:    mktInfo,
+		lotSize:       lotSize,
+		epochDuration: epochDuration,
+		orderRouter:   make(chan *orderUpdateSignal, 16),
+		book:          book.New(lotSize),
+		matcher:       matcher.New(),
+		swapper:       swap.NewSwapper(swapperCfg),
+		baseAsset:     baseAsset,
+		quoteAsset:    quoteAsset,
+		storage:       storage,
+	}, nil
+}
+
+func (m *Market) start() {
+	m.wg.Add(1)
+	go m.runEpochs()
+}
+
+func (m *Market) notifyNewOrder(order order.Order, action bookUpdateAction) {
+	// send to each receiver
+	for _, s := range m.orderFeeds {
+		select {
+		case <-m.ctx.Done():
+			return
+		case s <- &BookUpdateSignal{action, order}:
+		}
+	}
+}
+
+func (m *Market) runEpochs() {
+	// Startup epoch is the epoch after the epoch one second in the future.
+	dur := int64(m.epochDuration)
+	nextEpochIdx := 1 + (1+time.Now().Unix())%dur
+	nextEpoch := NewEpoch(nextEpochIdx, dur)
+	epochCycle := time.After(time.Until(nextEpoch.Start))
+
+	defer m.wg.Done()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.running = false
+			return
+
+		case <-epochCycle:
+			m.running = true
+			if m.epochQueue != nil {
+				go m.processEpoch(m.epochQueue)
+			}
+			m.epochQueue = nextEpoch
+
+			nextEpochIdx = m.epochQueue.Epoch + 1
+			nextEpoch = NewEpoch(nextEpochIdx, dur)
+			epochCycle = time.After(time.Until(nextEpoch.Start))
+
+		case s := <-m.orderRouter:
+			if m.epochQueue == nil /*!m.running*/ {
+				log.Debugf("Order %v received prior to market start.")
+				s.errChan <- ErrMarketNotRunning
+				continue
+			}
+
+			if ord := m.epochQueue.Orders[s.ord.ID()]; ord != nil {
+				// TODO: remove this check when we are sure the order router is behaving
+				log.Warnf("Received duplicate order %v!", ord)
+			}
+
+			// Validate the order. The order router must do it's own validation,
+			// but do a second validation before putting it on the queue.
+			if !m.validateOrder(s.ord) {
+				s.errChan <- ErrInvalidOrder
+				continue
+			}
+
+			// Insert the order into the epoch queue.
+			m.epochQueue.Orders[s.ord.ID()] = s.ord
+			s.errChan <- nil
```

Comment:
Note `errChan`.  Not necessary if the order router doesn't need to know, or if the order processing errors should get send back to the client via other means.

Reply:
Not sure yet. As another options, we could potentially send the message ID as part of the `orderUpdateSignal`. Errors could then be sent through the `AuthManager`.

Reply:
I'm not sure I have a clear picture of how that would work, so please see how the current PR works and decide if we need to implement error passing as you suggested.